# `House Rent Prediction`

![](https://tse2.mm.bing.net/th?id=OIP.dUAGp6e1tfdTFOzQzNiJ5wHaE7&pid=Api&P=0)

**This Project includes training the Machine Learning model to predict the House rent of a certain Loaction.**

**Steps included in this process :**
* Installing and Importing Libraries
* Loading the training and test dataset given in the folder data
* Doing Feature Engineering and Exploratory Data Analysis
* Scaling the Numeric Columns and Encoding the categorical Columns
* Splitting the Dataset into train,validation and test dataset 
* Training the model and finding the best algorithm with highest accuarcy

## `Installing libraries`

In [1]:
!pip install pandas numpy sklearn matplotlib  xgboost 

You should consider upgrading via the 'C:\Users\Aryan Sinha\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.


## `Importing useful libraries`

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## `Loading the Dataset and Performing EDA`

In [3]:
# Loading the train data
train_df = pd.read_csv('data/train.csv')
train_df.head()

id  type                    locality  \
0  ff8081815733a243015733b2876600a6  BHK2  Cauvery Colony,Koramangala   
1  ff8081815ee25e15015ee50004da2acd  BHK2                   Bellandur   
2  ff80818163f29c9e0163f46770b873e7  BHK3               Thiruvanmiyur   
3  ff80818164b68a700164b74b1b247a1d  BHK1                   Attiguppe   
4  ff80818163b1faf00163b4b9b8a163b1  BHK3                   Kodihalli   

    activation_date   latitude  longitude lease_type  gym  lift  \
0  21-04-2018 14:44  12.934471  77.634471     FAMILY    1     1   
1  09-10-2017 12:48  12.929557  77.672280     ANYONE    0     1   
2  12-06-2018 22:02  12.982870  80.262012     FAMILY    0     1   
3  20-07-2018 16:22  12.955991  77.531634     FAMILY    0     0   
4  31-05-2018 17:41  12.963903  77.649446     FAMILY    0     0   

   swimming_pool  ...  bathroom facing cup_board  floor  total_floor  \
0              1  ...         2      E         2      6           12   
1              0  ...         2     NE         2      3            4   
2              0  ...         3      E         3      1            5   
3              0  ...         1      E         1      1            2   
4              0  ...         3      E         4      0            0   

                                           amenities water_supply  \
0  {"LIFT":true,"GYM":true,"INTERNET":true,"AC":f...    CORP_BORE   
1  {"LIFT":true,"GYM":false,"INTERNET":false,"AC"...  CORPORATION   
2  {"LIFT":true,"GYM":false,"INTERNET":false,"AC"...    CORP_BORE   
3  {"LIFT":false,"GYM":false,"INTERNET":false,"AC...  CORPORATION   
4  {"LIFT":false,"GYM":false,"INTERNET":false,"AC...  CORPORATION   

   building_type  balconies   rent  
0             AP          2  40000  
1             AP          2  22000  
2             AP          3  28000  
3             IH          0   8000  
4             IH          1  45000  

[5 rows x 25 columns]

In [4]:
# Loading the test data as well
test_df = pd.read_csv('data/test.csv')
test_df.head()

id  type                       locality  \
0  ff8081815df539bc015df947ce976cca  BHK2                   Basavanagudi   
1  ff80818157288e9301572c05651853a6  BHK2                   Rajaji Nagar   
2  ff8081815f1afc58015f1b831fde166e  BHK1              Jeevan Bima Nagar   
3  ff8081816035b31901603f7922c046e0  BHK1            Jayanagar 1st Block   
4  ff8081815c3582f5015c38fc31aa73a4  BHK3  Whitefield Hope Farm Junction   

    activation_date   latitude  longitude lease_type  gym  lift  \
0  22-08-2017 09:00  12.941603  77.568156     FAMILY    0     1   
1  17-09-2017 16:33  12.998803  77.561887     ANYONE    0     0   
2  14-04-2018 17:27  12.966467  77.661063     ANYONE    0     0   
3  11-12-2017 13:15  12.941533  77.592606     ANYONE    0     0   
4  11-07-2017 16:44  12.971083  77.751625     ANYONE    1     1   

   swimming_pool  ...  property_age bathroom facing  cup_board  floor  \
0              0  ...             1        2      N          2      3   
1              0  ...             7        1      S          2      0   
2              0  ...            10        1      S          1      0   
3              0  ...             8        1      E          0      2   
4              1  ...             0        3      E          3      3   

   total_floor                                          amenities  \
0            3  {"LIFT":true,"GYM":false,"INTERNET":false,"AC"...   
1            3  {"LIFT":false,"GYM":false,"INTERNET":true,"AC"...   
2            1  {"LIFT":false,"GYM":false,"INTERNET":true,"AC"...   
3            2  {"LIFT":false,"GYM":false,"INTERNET":true,"AC"...   
4            4  {"LIFT":true,"GYM":true,"INTERNET":true,"AC":f...   

   water_supply  building_type  balconies  
0     CORP_BORE             AP          1  
1   CORPORATION             IF          0  
2   CORPORATION             IF          0  
3     CORP_BORE             IF          0  
4      BOREWELL             AP          1  

[5 rows x 24 columns]

In [5]:
train_df.columns

Index(['id', 'type', 'locality', 'activation_date', 'latitude', 'longitude',
       'lease_type', 'gym', 'lift', 'swimming_pool', 'negotiable',
       'furnishing', 'parking', 'property_size', 'property_age', 'bathroom',
       'facing', 'cup_board', 'floor', 'total_floor', 'amenities',
       'water_supply', 'building_type', 'balconies', 'rent'],
      dtype='object')

In [6]:
train_df.shape,test_df.shape

((20500, 25), (4500, 24))

In [7]:
train_df.isnull().sum()

id                 0
type               0
locality           0
activation_date    0
latitude           0
longitude          0
lease_type         0
gym                0
lift               0
swimming_pool      0
negotiable         0
furnishing         0
parking            0
property_size      0
property_age       0
bathroom           0
facing             0
cup_board          0
floor              0
total_floor        0
amenities          0
water_supply       0
building_type      0
balconies          0
rent               0
dtype: int64

In [8]:
test_df.isnull().sum()

id                 0
type               0
locality           0
activation_date    0
latitude           0
longitude          0
lease_type         0
gym                0
lift               0
swimming_pool      0
negotiable         0
furnishing         0
parking            0
property_size      0
property_age       0
bathroom           0
facing             0
cup_board          0
floor              0
total_floor        0
amenities          0
water_supply       0
building_type      0
balconies          0
dtype: int64

**So we do not have any null values in any column**

In [9]:
train_df.corr()

latitude  longitude       gym      lift  swimming_pool  \
latitude       1.000000   0.194742  0.035853  0.019632       0.034771   
longitude      0.194742   1.000000  0.245978  0.208302       0.229751   
gym            0.035853   0.245978  1.000000  0.645255       0.769686   
lift           0.019632   0.208302  0.645255  1.000000       0.564949   
swimming_pool  0.034771   0.229751  0.769686  0.564949       1.000000   
negotiable    -0.009285  -0.024382  0.008972  0.011361       0.010066   
property_size  0.006797   0.075645  0.290462  0.308248       0.284686   
property_age   0.030060  -0.136155 -0.179643 -0.181409      -0.167475   
bathroom      -0.007669   0.095956  0.327032  0.366707       0.320946   
cup_board     -0.003166  -0.004139 -0.002625 -0.003867      -0.002198   
floor          0.004313   0.152689  0.392266  0.405142       0.434442   
total_floor   -0.000597   0.201782  0.526721  0.530908       0.574944   
balconies      0.049549   0.164898  0.415512  0.525769       0.383183   
rent           0.033130   0.163864  0.432081  0.462606       0.434776   

               negotiable  property_size  property_age  bathroom  cup_board  \
latitude        -0.009285       0.006797      0.030060 -0.007669  -0.003166   
longitude       -0.024382       0.075645     -0.136155  0.095956  -0.004139   
gym              0.008972       0.290462     -0.179643  0.327032  -0.002625   
lift             0.011361       0.308248     -0.181409  0.366707  -0.003867   
swimming_pool    0.010066       0.284686     -0.167475  0.320946  -0.002198   
negotiable       1.000000       0.053598      0.038837  0.059541   0.004701   
property_size    0.053598       1.000000     -0.008227  0.594711  -0.002578   
property_age     0.038837      -0.008227      1.000000 -0.047173   0.004564   
bathroom         0.059541       0.594711     -0.047173  1.000000  -0.006370   
cup_board        0.004701      -0.002578      0.004564 -0.006370   1.000000   
floor           -0.012456       0.148342     -0.192410  0.171226  -0.002205   
total_floor     -0.002173       0.226368     -0.231279  0.265728  -0.003157   
balconies        0.027089       0.379853     -0.108382  0.448801  -0.006551   
rent             0.062430       0.588425     -0.010786  0.677516  -0.001247   

                  floor  total_floor  balconies      rent  
latitude       0.004313    -0.000597   0.049549  0.033130  
longitude      0.152689     0.201782   0.164898  0.163864  
gym            0.392266     0.526721   0.415512  0.432081  
lift           0.405142     0.530908   0.525769  0.462606  
swimming_pool  0.434442     0.574944   0.383183  0.434776  
negotiable    -0.012456    -0.002173   0.027089  0.062430  
property_size  0.148342     0.226368   0.379853  0.588425  
property_age  -0.192410    -0.231279  -0.108382 -0.010786  
bathroom       0.171226     0.265728   0.448801  0.677516  
cup_board     -0.002205    -0.003157  -0.006551 -0.001247  
floor          1.000000     0.710419   0.327213  0.306769  
total_floor    0.710419     1.000000   0.361790  0.437437  
balconies      0.327213     0.361790   1.000000  0.472405  
rent           0.306769     0.437437   0.472405  1.000000

There is no need of columns like id, activation_date, amenities as they are not playing any role in price prediction.

Also floor and total_floor have high correlation (about 71%) so we can drop any of them (Let's drop total_floor column).

In [10]:
train_df.drop(['id','activation_date','amenities','total_floor'],axis=1,inplace=True)
test_df.drop(['id','activation_date','amenities','total_floor'],axis=1,inplace=True)

In [11]:
train_df.shape,test_df.shape

((20500, 21), (4500, 20))

In [12]:
train_df.describe()

latitude     longitude           gym          lift  swimming_pool  \
count  20500.000000  20500.000000  20500.000000  20500.000000   20500.000000   
mean      12.945751     77.637280      0.226732      0.361463       0.174927   
std        0.029594      0.110586      0.418728      0.480436       0.379914   
min       12.900004     77.500072      0.000000      0.000000       0.000000   
25%       12.918188     77.571956      0.000000      0.000000       0.000000   
50%       12.943800     77.633491      0.000000      0.000000       0.000000   
75%       12.970952     77.695760      0.000000      1.000000       0.000000   
max       12.999999     80.266346      1.000000      1.000000       1.000000   

         negotiable  property_size  property_age      bathroom     cup_board  \
count  20500.000000   20500.000000  20500.000000  20500.000000  20500.000000   
mean       0.711512    1061.829463      5.770634      1.858146      4.553512   
std        0.453070     591.419631      6.350271      0.710822    335.416267   
min        0.000000       1.000000     -1.000000      1.000000      0.000000   
25%        0.000000     700.000000      1.000000      1.000000      1.000000   
50%        1.000000    1047.000000      5.000000      2.000000      2.000000   
75%        1.000000    1281.000000     10.000000      2.000000      3.000000   
max        1.000000   50000.000000    400.000000     21.000000  48026.000000   

              floor     balconies          rent  
count  20500.000000  20500.000000  20500.000000  
mean       1.873561      1.148976  19053.602195  
std        2.184927      0.993835   8433.949300  
min        0.000000      0.000000   8000.000000  
25%        1.000000      0.000000  12500.000000  
50%        1.000000      1.000000  17500.000000  
75%        2.000000      2.000000  24000.000000  
max       25.000000     13.000000  50000.000000

**With the help of above description we can see that there are some outliers like in property_size, property_age, bathroom, cup_board & balconies as in all of them 75% Quartile value is much less than the maximum value and there are some values also which are not even possible (for eg. cup_board = 48026)**

In [13]:
test_df.describe()

latitude    longitude          gym         lift  swimming_pool  \
count  4500.000000  4500.000000  4500.000000  4500.000000    4500.000000   
mean     12.946577    77.638205     0.223111     0.366889       0.173111   
std       0.029694     0.116819     0.416379     0.482009       0.378385   
min      12.900007    77.500096     0.000000     0.000000       0.000000   
25%      12.919406    77.572006     0.000000     0.000000       0.000000   
50%      12.944424    77.632945     0.000000     0.000000       0.000000   
75%      12.971717    77.697231     0.000000     1.000000       0.000000   
max      12.999929    80.257515     1.000000     1.000000       1.000000   

        negotiable  property_size  property_age     bathroom    cup_board  \
count  4500.000000    4500.000000   4500.000000  4500.000000  4500.000000   
mean      0.719333    1072.577778      5.827111     1.874667     2.244222   
std       0.449375     590.043133     15.800894     0.950581     1.610327   
min       0.000000       1.000000     -1.000000     1.000000     0.000000   
25%       0.000000     700.000000      1.000000     1.000000     1.000000   
50%       1.000000    1064.000000      5.000000     2.000000     2.000000   
75%       1.000000    1300.000000      9.000000     2.000000     3.000000   
max       1.000000   20000.000000    999.000000    34.000000    30.000000   

             floor    balconies  
count  4500.000000  4500.000000  
mean      1.918222     1.166444  
std       2.201149     1.074915  
min       0.000000     0.000000  
25%       1.000000     0.000000  
50%       1.000000     1.000000  
75%       3.000000     2.000000  
max      21.000000    22.000000

Similar case we see in test dataset

**Let's resolve the issue**

In [14]:
## Property_size

train_df[train_df.property_size>8000]

type            locality   latitude  longitude lease_type  gym  lift  \
549    BHK3  Kumaraswamy Layout  12.908044  77.561586     FAMILY    0     1   
5144   BHK3        Basavanagudi  12.936154  77.568650     ANYONE    0     0   
6075   BHK2         Munnekollal  12.950302  77.710188     ANYONE    0     0   
16379  BHK3          HSR Layout  12.918079  77.637825     FAMILY    0     0   

       swimming_pool  negotiable       furnishing  ... property_size  \
549                0           1  FULLY_FURNISHED  ...         13000   
5144               0           0  FULLY_FURNISHED  ...         10200   
6075               0           1   SEMI_FURNISHED  ...         50000   
16379              0           1   SEMI_FURNISHED  ...         12000   

       property_age  bathroom  facing cup_board  floor  water_supply  \
549               5         2       N         3      1      BOREWELL   
5144              3         3      NE         5      0   CORPORATION   
6075              1         2       N         2      2     CORP_BORE   
16379             1         3       N         2      1     CORP_BORE   

      building_type balconies   rent  
549              AP         1  17000  
5144             IH         1  40000  
6075             IF         0  16000  
16379            IH         1  18000  

[4 rows x 21 columns]

It's not possible to have such a huge size with equivalent price that of others houses with lesser size.

So lets drop these rows.

In [15]:
train_df = train_df[train_df.property_size<8000]

In [16]:
## property_age

train_df[train_df.property_age>50]

type         locality   latitude  longitude lease_type  gym  lift  \
2956   BHK2    Shivaji Nagar  12.986714  77.612233     ANYONE    0     0   
10493  BHK2        Jayanagar  12.930946  77.582201     ANYONE    0     0   
12769  BHK2       Whitefield  12.955611  77.722438     ANYONE    0     0   
17913  BHK2  Manjunath Nagar  12.997467  77.549205     ANYONE    0     0   
20132  BHK2    Srinivasnagar  12.938325  77.552962     FAMILY    0     0   

       swimming_pool  negotiable      furnishing  ... property_size  \
2956               0           1   NOT_FURNISHED  ...          1800   
10493              0           1   NOT_FURNISHED  ...          1800   
12769              0           1   NOT_FURNISHED  ...          1200   
17913              0           1  SEMI_FURNISHED  ...           800   
20132              0           0  SEMI_FURNISHED  ...           600   

       property_age  bathroom  facing cup_board  floor  water_supply  \
2956             80         1       N         0      0   CORPORATION   
10493            60         1      SW         0      1   CORPORATION   
12769           400         1       E         0      0      BOREWELL   
17913           100         1       W         2      1   CORPORATION   
20132           100         1       N         2      2     CORP_BORE   

      building_type balconies   rent  
2956             IF         0  12000  
10493            IF         0  25000  
12769            IF         0  25000  
17913            IF         0  10000  
20132            IF         0   9000  

[5 rows x 21 columns]

Dropping these rows as this might not possible to have a house with more than 50 years age.

This may be possible but I am not going to consider it as 75% quartile (means 75% of houses in our dataset) is only 10.

In [17]:
train_df = train_df[train_df.property_age<=50]

In [18]:
## Bathroom

train_df[train_df.bathroom>6]

type  locality   latitude  longitude lease_type  gym  lift  \
2431       BHK3   Ejipura  12.945015  77.628737   BACHELOR    0     0   
18138  BHK4PLUS  JP Nagar  12.907575  77.573134     ANYONE    0     0   

       swimming_pool  negotiable       furnishing  ... property_size  \
2431               0           1  FULLY_FURNISHED  ...          1100   
18138              0           0  FULLY_FURNISHED  ...          2500   

       property_age  bathroom  facing cup_board  floor  water_supply  \
2431              5        21      NE         3      0     CORP_BORE   
18138            10         7       E        20      1     CORP_BORE   

      building_type balconies   rent  
2431             IH        13  27000  
18138            IF         1  46000  

[2 rows x 21 columns]

So it's clear from the above that there no such house exist with such a small size and lots of bathroom (20 bathrooms in 1100 property size).

In [19]:
train_df = train_df[train_df.bathroom<=6]

In [20]:
## cup_board

train_df[train_df.cup_board>=15]

type                                           locality   latitude  \
946    BHK3                                      Doddanekkundi  12.971917   
1934   BHK3                             Devasthanagalu,Varthur  12.939309   
2800   BHK3                                       Mahadevapura  12.980592   
7568   BHK2                                        Brookefield  12.965860   
7650   BHK2                      Adarsh Palm Retreat,Bellandur  12.920331   
9809   BHK2                 Kasavanahalli Village,Marathahalli  12.946938   
10276  BHK3                                      Kaggadasapura  12.981847   
12889  BHK2  LB Shastri Nagar, Kaggadasapura, Bengaluru, Ka...  12.967071   
14580  BHK3                                      Wilson Garden  12.943940   
15861  BHK3                                       Marathahalli  12.959570   
17624  BHK3                                       Bommanahalli  12.900284   
19153  BHK2                                       Basavanagudi  12.932136   
19327  BHK2                          Ambedkar Nagar,Whitefield  12.982643   

       longitude lease_type  gym  lift  swimming_pool  negotiable  \
946    77.704298     ANYONE    1     1              0           1   
1934   77.736541     FAMILY    1     1              1           1   
2800   77.692662     FAMILY    0     1              1           0   
7568   77.720689     FAMILY    1     1              0           1   
7650   77.688482     FAMILY    1     1              1           0   
9809   77.702844     ANYONE    1     1              0           1   
10276  77.666921     ANYONE    0     0              0           1   
12889  77.670486     FAMILY    0     0              0           1   
14580  77.595673     FAMILY    0     0              0           1   
15861  77.710244     FAMILY    1     1              1           1   
17624  77.633031    COMPANY    0     0              0           1   
19153  77.566706     FAMILY    0     0              0           1   
19327  77.755187     ANYONE    1     1              1           1   

            furnishing  ... property_size  property_age  bathroom  facing  \
946     SEMI_FURNISHED  ...          1400             1         2       W   
1934    SEMI_FURNISHED  ...          1560             1         3       E   
2800    SEMI_FURNISHED  ...          2000             1         3       E   
7568    SEMI_FURNISHED  ...          1400             4         2       W   
7650    SEMI_FURNISHED  ...          1805             2         2       W   
9809   FULLY_FURNISHED  ...          1300             2         2       E   
10276   SEMI_FURNISHED  ...          1500             7         2       W   
12889   SEMI_FURNISHED  ...          1200             5         2       E   
14580   SEMI_FURNISHED  ...          1785            10         3       N   
15861   SEMI_FURNISHED  ...          1505            10         3       N   
17624   SEMI_FURNISHED  ...          2000            10         3       E   
19153   SEMI_FURNISHED  ...           700            10         1       N   
19327  FULLY_FURNISHED  ...          1450             1         2       S   

      cup_board  floor  water_supply building_type balconies   rent  
946          15      3     CORP_BORE            AP         1  25000  
1934         20      4      BOREWELL            AP         2  20000  
2800         15      1     CORP_BORE            AP         2  39000  
7568         20      5     CORP_BORE            AP         1  21000  
7650         15      0     CORP_BORE            AP         2  35000  
9809         15      1     CORP_BORE            AP         2  25000  
10276        16      0     CORP_BORE            AP         2  25000  
12889        15      3     CORP_BORE            AP         1  22000  
14580        16      2   CORPORATION            AP         3  33500  
15861        15      2     CORP_BORE            AP         2  24500  
17624        40      3      BOREWELL            AP         2  40000  
19153     48026      1     CORP_BORE      

It's not possible to have more than 15 cupboards.

There is one outlier as well with 48026 cup_borads in property_size of 700 sq.ft.

In [21]:
train_df = train_df[train_df.cup_board<15]

In [22]:
## Balconies

train_df = train_df[train_df.balconies<=5]

In [23]:
## Floor

train_df[train_df.floor>18]

type           locality   latitude  longitude lease_type  gym  lift  \
1212   BHK2       Bommanahalli  12.901398  77.628912     FAMILY    1     1   
1846   BHK3       Bommanahalli  12.901398  77.628987     ANYONE    1     1   
5897   BHK3  Halanayakanahalli  12.910441  77.682649     FAMILY    1     1   
6028   BHK2       Mahadevapura  12.991456  77.702748     FAMILY    1     1   
7018   BHK3       Mahadevapura  12.991464  77.702998     FAMILY    1     1   
9252   BHK2     Kaikondrahalli  12.912897  77.680961     FAMILY    1     1   
9431   BHK2     Kaikondrahalli  12.912897  77.680961     FAMILY    1     1   
9893   BHK3           parkview  12.985015  77.752137     ANYONE    1     1   
12487  BHK3       Bommanahalli  12.901747  77.631216     ANYONE    1     1   
15181  BHK3  Uttarahalli Hobli  12.908861  77.540916     FAMILY    1     1   
17238  BHK2       Bommanahalli  12.901403  77.628901     ANYONE    0     0   

       swimming_pool  negotiable       furnishing  ... property_size  \
1212               1           1   SEMI_FURNISHED  ...          1345   
1846               1           1   SEMI_FURNISHED  ...          1590   
5897               1           1   SEMI_FURNISHED  ...          1499   
6028               1           1   SEMI_FURNISHED  ...          1290   
7018               1           0   SEMI_FURNISHED  ...          1640   
9252               1           1   SEMI_FURNISHED  ...          1200   
9431               0           0   SEMI_FURNISHED  ...          1260   
9893               1           1  FULLY_FURNISHED  ...          1840   
12487              1           1   SEMI_FURNISHED  ...          1590   
15181              1           1   SEMI_FURNISHED  ...          1900   
17238              0           1  FULLY_FURNISHED  ...          1325   

       property_age  bathroom  facing cup_board  floor  water_supply  \
1212              4         2       S         4     25     CORP_BORE   
1846              2         2       N         3     20     CORP_BORE   
5897              0         2       E         3     19     CORP_BORE   
6028              6         2       W         2     19     CORP_BORE   
7018             10         3       E         3     19     CORP_BORE   
9252              5         2      SW         2     22     CORP_BORE   
9431              0         2       N         2     20     CORP_BORE   
9893              1         3       E         3     19     CORP_BORE   
12487             0         2       E         3     20     CORP_BORE   
15181             2         3       E         8     19   CORPORATION   
17238             5         2       N         0     19   CORPORATION   

      building_type balconies   rent  
1212             AP         2  34000  
1846             AP         2  36000  
5897             AP         2  38000  
6028             AP         3  32000  
7018             AP         3  38000  
9252             AP         4  30000  
9431             AP         2  33000  
9893             AP         2  40000  
12487            AP         2  45000  
15181            AP         3  25000  
17238            AP         2  39000  

[11 rows x 21 columns]

As 75% of houses are having max of 2 floors so I am going to intentionally drop these houses data having floor greater than 18 as it will affect my accuracy.

In [24]:
train_df = train_df[train_df.floor<=18]

In [25]:
train_df.shape,test_df.shape

((20458, 21), (4500, 20))

**So finally we are done with outliers**

In [26]:
train_df.describe()

latitude     longitude           gym         lift  swimming_pool  \
count  20458.000000  20458.000000  20458.000000  20458.00000   20458.000000   
mean      12.945756     77.637262      0.226268      0.36113       0.174553   
std        0.029593      0.110656      0.418425      0.48034       0.379593   
min       12.900004     77.500072      0.000000      0.00000       0.000000   
25%       12.918209     77.571914      0.000000      0.00000       0.000000   
50%       12.943794     77.633484      0.000000      0.00000       0.000000   
75%       12.970974     77.695715      0.000000      1.00000       0.000000   
max       12.999999     80.266346      1.000000      1.00000       1.000000   

         negotiable  property_size  property_age      bathroom     cup_board  \
count  20458.000000   20458.000000  20458.000000  20458.000000  20458.000000   
mean       0.711409    1056.900479      5.738195      1.856242      2.199971   
std        0.453119     463.729192      5.612853      0.696790      1.483298   
min        0.000000       1.000000     -1.000000      1.000000      0.000000   
25%        0.000000     700.000000      1.000000      1.000000      1.000000   
50%        1.000000    1045.000000      5.000000      2.000000      2.000000   
75%        1.000000    1280.000000     10.000000      2.000000      3.000000   
max        1.000000    6040.000000     50.000000      6.000000     13.000000   

              floor     balconies          rent  
count  20458.000000  20458.000000  20458.000000  
mean       1.863085      1.145469  19034.526591  
std        2.142108      0.982197   8417.342779  
min        0.000000      0.000000   8000.000000  
25%        1.000000      0.000000  12500.000000  
50%        1.000000      1.000000  17500.000000  
75%        2.000000      2.000000  24000.000000  
max       18.000000      5.000000  50000.000000

In [27]:
test_df.describe()

latitude    longitude          gym         lift  swimming_pool  \
count  4500.000000  4500.000000  4500.000000  4500.000000    4500.000000   
mean     12.946577    77.638205     0.223111     0.366889       0.173111   
std       0.029694     0.116819     0.416379     0.482009       0.378385   
min      12.900007    77.500096     0.000000     0.000000       0.000000   
25%      12.919406    77.572006     0.000000     0.000000       0.000000   
50%      12.944424    77.632945     0.000000     0.000000       0.000000   
75%      12.971717    77.697231     0.000000     1.000000       0.000000   
max      12.999929    80.257515     1.000000     1.000000       1.000000   

        negotiable  property_size  property_age     bathroom    cup_board  \
count  4500.000000    4500.000000   4500.000000  4500.000000  4500.000000   
mean      0.719333    1072.577778      5.827111     1.874667     2.244222   
std       0.449375     590.043133     15.800894     0.950581     1.610327   
min       0.000000       1.000000     -1.000000     1.000000     0.000000   
25%       0.000000     700.000000      1.000000     1.000000     1.000000   
50%       1.000000    1064.000000      5.000000     2.000000     2.000000   
75%       1.000000    1300.000000      9.000000     2.000000     3.000000   
max       1.000000   20000.000000    999.000000    34.000000    30.000000   

             floor    balconies  
count  4500.000000  4500.000000  
mean      1.918222     1.166444  
std       2.201149     1.074915  
min       0.000000     0.000000  
25%       1.000000     0.000000  
50%       1.000000     1.000000  
75%       3.000000     2.000000  
max      21.000000    22.000000

In [28]:
train_df.head(2)

type                    locality   latitude  longitude lease_type  gym  \
0  BHK2  Cauvery Colony,Koramangala  12.934471  77.634471     FAMILY    1   
1  BHK2                   Bellandur  12.929557  77.672280     ANYONE    0   

   lift  swimming_pool  negotiable      furnishing  ... property_size  \
0     1              1           0  SEMI_FURNISHED  ...          1250   
1     1              0           1  SEMI_FURNISHED  ...          1400   

   property_age  bathroom  facing cup_board  floor  water_supply  \
0            25         2       E         2      6     CORP_BORE   
1             4         2      NE         2      3   CORPORATION   

  building_type balconies   rent  
0            AP         2  40000  
1            AP         2  22000  

[2 rows x 21 columns]

#### Dealing with BHK Value

In [29]:
train_df.type.value_counts()

BHK2        11660
BHK3         4385
BHK1         3600
RK1           553
BHK4          230
BHK4PLUS       30
Name: type, dtype: int64

In [30]:
test_df.type.value_counts()

BHK2        2524
BHK3         990
BHK1         808
RK1          105
BHK4          66
BHK4PLUS       7
Name: type, dtype: int64

As there are very few houses with more than 4BHK and so lets make them 5

In [31]:
# Now making a column 'bhk' which will be numeric and removing 'type' column

def find_bhk(s):
    if(s=='BHK4PLUS'):
        return 5
    return int(s[len(s)-1])

# print(type(find_bhk('RK4')))

train_df.type = train_df.type.apply(find_bhk)
test_df.type = test_df.type.apply(find_bhk)

train_df.rename(columns={'type':'bhk'},inplace=True)

In [32]:
train_df

bhk                    locality   latitude  longitude lease_type  gym  \
0        2  Cauvery Colony,Koramangala  12.934471  77.634471     FAMILY    1   
1        2                   Bellandur  12.929557  77.672280     ANYONE    0   
2        3               Thiruvanmiyur  12.982870  80.262012     FAMILY    0   
3        1                   Attiguppe  12.955991  77.531634     FAMILY    0   
4        3                   Kodihalli  12.963903  77.649446     FAMILY    0   
...    ...                         ...        ...        ...        ...  ...   
20495    3          Gattigere,RR Nagar  12.925380  77.524397     ANYONE    1   
20496    2       Koramangala 4th Block  12.931072  77.632361     ANYONE    0   
20497    2          Kumaraswamy Layout  12.901800  77.566597     FAMILY    1   
20498    2                 Indiranagar  12.965502  77.647694     FAMILY    0   
20499    1           Malleshwaram West  12.998645  77.574336     ANYONE    0   

       lift  swimming_pool  negotiable      furnishing  ... property_size  \
0         1              1           0  SEMI_FURNISHED  ...          1250   
1         1              0           1  SEMI_FURNISHED  ...          1400   
2         1              0           0  SEMI_FURNISHED  ...          1350   
3         0              0           1  SEMI_FURNISHED  ...           600   
4         0              0           1  SEMI_FURNISHED  ...          1500   
...     ...            ...         ...             ...  ...           ...   
20495     1              1           1  SEMI_FURNISHED  ...          1650   
20496     0              0           1  SEMI_FURNISHED  ...          1000   
20497     1              0           1  SEMI_FURNISHED  ...          1000   
20498     1              0           0  SEMI_FURNISHED  ...          1000   
20499     0              0           1  SEMI_FURNISHED  ...           500   

       property_age  bathroom  facing cup_board  floor  water_supply  \
0                25         2       E         2      6     CORP_BORE   
1                 4         2      NE         2      3   CORPORATION   
2                 6         3       E         3      1     CORP_BORE   
3                 3         1       E         1      1   CORPORATION   
4                15         3       E         4      0   CORPORATION   
...             ...       ...     ...       ...    ...           ...   
20495             1         3       E         3      4      BOREWELL   
20496            10         2       N         2      1   CORPORATION   
20497             0         2       W         2      5      BOREWELL   
20498             5         2       E         2      2     CORP_BORE   
20499             3         1       E         1      1   CORPORATION   

      building_type balconies   rent  
0                AP         2  40000  
1                AP         2  22000  
2                AP         3  28000  
3                IH         0   8000  
4                IH         1  45000  
...             ...       ...    ...  
20495            AP         2  25000  
20496            IH         2  30000  
20497            AP         2  16000  
20498            AP         3  30000  
20499            IH         1   8500  

[20458 rows x 21 columns]

In [33]:
train_df.columns

Index(['bhk', 'locality', 'latitude', 'longitude', 'lease_type', 'gym', 'lift',
       'swimming_pool', 'negotiable', 'furnishing', 'parking', 'property_size',
       'property_age', 'bathroom', 'facing', 'cup_board', 'floor',
       'water_supply', 'building_type', 'balconies', 'rent'],
      dtype='object')

In [34]:
train_df.bhk.unique()

array([2, 3, 1, 4, 5], dtype=int64)

In [35]:
test_df.rename(columns={'type':'bhk'},inplace=True)
test_df.head(1)

bhk      locality   latitude  longitude lease_type  gym  lift  \
0    2  Basavanagudi  12.941603  77.568156     FAMILY    0     1   

   swimming_pool  negotiable      furnishing parking  property_size  \
0              0           0  SEMI_FURNISHED    BOTH           1300   

   property_age  bathroom facing  cup_board  floor water_supply building_type  \
0             1         2      N          2      3    CORP_BORE            AP   

   balconies  
0          1

#### Locality

In [36]:
train_df.locality = train_df.locality.apply(lambda x: x.strip())
test_df.locality = test_df.locality.apply(lambda x: x.strip())

location_stats_train = train_df['locality'].value_counts(ascending=False)
location_stats_test = test_df['locality'].value_counts(ascending=False)

print(location_stats_train)
# print(location_stats_test)

Whitefield                                                          907
HSR Layout                                                          683
Banashankari                                                        644
Marathahalli                                                        473
Bellandur                                                           439
                                                                   ... 
Bommenahalli                                                          1
Chikkalasandra, Bengaluru, Karnataka, India                           1
Koramangala 3 Block, Bengaluru, Karnataka, India                      1
Kumara Park West                                                      1
Mahadevapura bus stop, Mahadevapura, Bengaluru, Karnataka, India      1
Name: locality, Length: 1935, dtype: int64


In [37]:
location_stats_train.values.sum()

20458

In [38]:
len(location_stats_train[location_stats_train>50]),len(location_stats_test[location_stats_test>50])

(67, 19)

In [39]:
len(location_stats_train)

1935

In [40]:
## Dimensionality Reduction

location_stats_less_than_50_train = location_stats_train[location_stats_train<=50]
location_stats_less_than_50_test = location_stats_test[location_stats_test<=50]

location_stats_less_than_50_train

Attiguppe                                                           50
Thubarahalli,Whitefield                                             50
J P Nagar                                                           48
Murugeshpalya                                                       47
Jogupalya                                                           46
                                                                    ..
Bommenahalli                                                         1
Chikkalasandra, Bengaluru, Karnataka, India                          1
Koramangala 3 Block, Bengaluru, Karnataka, India                     1
Kumara Park West                                                     1
Mahadevapura bus stop, Mahadevapura, Bengaluru, Karnataka, India     1
Name: locality, Length: 1868, dtype: int64

In [41]:
location_stats_less_than_50_test

Kasavanahalli                                                                                                                                    50
Nagarbhavi                                                                                                                                       46
Rajaji Nagar                                                                                                                                     45
Ejipura                                                                                                                                          45
Basaveshwar Nagar                                                                                                                                43
                                                                                                                                                 ..
Venkateshwara Layout                                                                                            

In [42]:
len(train_df.locality.unique())

1935

In [43]:
train_df.locality = train_df.locality.apply(lambda x: 'other' if x in (location_stats_less_than_50_train & location_stats_less_than_50_test) else x)
test_df.locality = test_df.locality.apply(lambda x: 'other' if x in (location_stats_less_than_50_train & location_stats_less_than_50_test) else x)

len(train_df.locality.unique())

20

In [44]:
len(test_df.locality.unique())

20

In [45]:
train_df.locality.unique()

array(['other', 'Bellandur', 'Banashankari', 'Whitefield', 'RR Nagar',
       'BTM 2nd Stage', 'Munnekollal', 'Bommanahalli', 'Mahadevapura',
       'Brookefield', 'Jayanagar', 'Kaggadasapura', 'BTM Layout',
       'Vijaya Nagar', 'Marathahalli', 'HSR Layout', 'Koramangala',
       'Basavanagudi', 'JP Nagar', 'Varthur'], dtype=object)

In [46]:
test_df.locality.unique()

array(['Basavanagudi', 'other', 'Bellandur', 'HSR Layout',
       'Kaggadasapura', 'JP Nagar', 'Whitefield', 'BTM 2nd Stage',
       'RR Nagar', 'Jayanagar', 'Koramangala', 'Marathahalli',
       'Banashankari', 'Munnekollal', 'Mahadevapura', 'Bommanahalli',
       'Vijaya Nagar', 'Brookefield', 'BTM Layout', 'Varthur'],
      dtype=object)

In [47]:
train_df.head(2)

bhk   locality   latitude  longitude lease_type  gym  lift  swimming_pool  \
0    2      other  12.934471  77.634471     FAMILY    1     1              1   
1    2  Bellandur  12.929557  77.672280     ANYONE    0     1              0   

   negotiable      furnishing  ... property_size  property_age  bathroom  \
0           0  SEMI_FURNISHED  ...          1250            25         2   
1           1  SEMI_FURNISHED  ...          1400             4         2   

   facing cup_board  floor  water_supply building_type balconies   rent  
0       E         2      6     CORP_BORE            AP         2  40000  
1      NE         2      3   CORPORATION            AP         2  22000  

[2 rows x 21 columns]

In [48]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20458 entries, 0 to 20499
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   bhk            20458 non-null  int64  
 1   locality       20458 non-null  object 
 2   latitude       20458 non-null  float64
 3   longitude      20458 non-null  float64
 4   lease_type     20458 non-null  object 
 5   gym            20458 non-null  int64  
 6   lift           20458 non-null  int64  
 7   swimming_pool  20458 non-null  int64  
 8   negotiable     20458 non-null  int64  
 9   furnishing     20458 non-null  object 
 10  parking        20458 non-null  object 
 11  property_size  20458 non-null  int64  
 12  property_age   20458 non-null  int64  
 13  bathroom       20458 non-null  int64  
 14  facing         20458 non-null  object 
 15  cup_board      20458 non-null  int64  
 16  floor          20458 non-null  int64  
 17  water_supply   20458 non-null  object 
 18  buildi

**So Thats whole EDA and Data Wrangling**

In [49]:
train_df.head(2)

bhk   locality   latitude  longitude lease_type  gym  lift  swimming_pool  \
0    2      other  12.934471  77.634471     FAMILY    1     1              1   
1    2  Bellandur  12.929557  77.672280     ANYONE    0     1              0   

   negotiable      furnishing  ... property_size  property_age  bathroom  \
0           0  SEMI_FURNISHED  ...          1250            25         2   
1           1  SEMI_FURNISHED  ...          1400             4         2   

   facing cup_board  floor  water_supply building_type balconies   rent  
0       E         2      6     CORP_BORE            AP         2  40000  
1      NE         2      3   CORPORATION            AP         2  22000  

[2 rows x 21 columns]

In [50]:
X = train_df.drop(['rent'],axis=1)
y = train_df['rent']

In [51]:
from sklearn.model_selection import train_test_split
X_train,X_val,y_train,y_val = train_test_split(X,y,test_size=0.20,random_state=42)

In [52]:
X_train.head()

bhk      locality   latitude  longitude lease_type  gym  lift  \
4318     1     Bellandur  12.930895  77.675724     ANYONE    1     1   
19872    2  Banashankari  12.915945  77.545990     ANYONE    1     1   
7401     2         other  12.971275  77.767599     FAMILY    0     0   
16273    3         other  12.939414  77.695203     ANYONE    1     1   
6495     2         other  12.974279  77.693331     FAMILY    0     1   

       swimming_pool  negotiable       furnishing      parking  property_size  \
4318               1           1  FULLY_FURNISHED         BOTH            600   
19872              1           1   SEMI_FURNISHED         BOTH           1260   
7401               0           1   SEMI_FURNISHED  TWO_WHEELER            950   
16273              1           1   SEMI_FURNISHED         BOTH           1306   
6495               0           0   SEMI_FURNISHED         BOTH           1100   

       property_age  bathroom facing  cup_board  floor water_supply  \
4318              0         1      W          2      3    CORP_BORE   
19872            10         2      N          2      1    CORP_BORE   
7401              5         2      E          2      0     BOREWELL   
16273             0         2      E          4      3     BOREWELL   
6495              5         2      E          2      1    CORP_BORE   

      building_type  balconies  
4318             AP          3  
19872            AP          2  
7401             IH          2  
16273            AP          3  
6495             AP          1

In [53]:
X_train.shape,X_val.shape

((16366, 20), (4092, 20))

In [54]:
X_test = test_df
X_test.shape

(4500, 20)

## `Creating the Pipeline`

In [55]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import r2_score,mean_absolute_error

## `Importing Regression Models`

In [56]:
from sklearn.linear_model import LinearRegression,Ridge,Lasso
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor,GradientBoostingRegressor,AdaBoostRegressor,ExtraTreesRegressor
from sklearn.svm import SVR
from xgboost import XGBRegressor

**Let's try different different Regression Models now ..**

## `Linear Regression`

In [57]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = LinearRegression()

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.7134824099249683
Mean Absolute error 3342.9392670945754


## `Ridge Regression`

In [58]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = Ridge(alpha=10)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.7101981149924591
Mean Absolute error 3359.4598665648964


## `LASSO Regression`

In [59]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = Lasso(alpha=10)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.7075471260349362
Mean Absolute error 3369.197005888183


## `KNN Regressor`

In [60]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = KNeighborsRegressor(n_neighbors=5)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.49044173296086635
Mean Absolute error 4366.979032258065


## `Decision Tree Regressor`

In [61]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = DecisionTreeRegressor(max_depth=9)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.722736570428383
Mean Absolute error 3189.883978653729


## `Support Vector Machine Regressor`

In [62]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = SVR(kernel='rbf',C=10000,epsilon=0.1)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.7247671875031453
Mean Absolute error 3194.249933867378


## `Random Forest Regressor`

In [63]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = RandomForestRegressor(n_jobs=-1, 
                               random_state=42, 
                               n_estimators=90,
                               max_features=50,
                               max_depth=40, 
                               max_leaf_nodes=2**20)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.8094283155348099
Mean Absolute error 2634.2815493646144


## `Extra Tree Regressor`

In [64]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = ExtraTreesRegressor(n_estimators=100,
                              random_state=3,
                              max_samples=0.5,
                              max_features=0.75,
                              max_depth=15)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.7713419720155587
Mean Absolute error 2943.462982321428


## `AdaBoost Regressor`

In [65]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = AdaBoostRegressor(n_estimators=15,learning_rate=1.0)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.629199534349945
Mean Absolute error 4145.583193251969


## `Gradient Boosting Regressor`

In [66]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = GradientBoostingRegressor(n_estimators=500)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.8251900760378057
Mean Absolute error 2576.13461706702


## `XGBoost Regressor`

In [67]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = XGBRegressor(n_estimators=45,max_depth=5,learning_rate=0.5)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.8194005411603165
Mean Absolute error 2607.109561864232


**Trying two more different types of Regressor**

In [68]:
from sklearn.ensemble import VotingRegressor,StackingRegressor

## `Voting Regressor`

In [69]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

rf = RandomForestRegressor(n_estimators=350,random_state=3,max_samples=0.5,max_features=0.75,max_depth=15)
gbdt = GradientBoostingRegressor(n_estimators=100,max_features=0.5)
xgb = XGBRegressor(n_estimators=25,learning_rate=0.3,max_depth=5)
et = ExtraTreesRegressor(n_estimators=100,random_state=3,max_samples=0.5,max_features=0.75,max_depth=10)

step2 = VotingRegressor([('rf', rf), ('gbdt', gbdt), ('xgb',xgb), ('et',et)],weights=[5,1,1,1])

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.8087437324009927
Mean Absolute error 2680.651095016353


## `Stacking Regressor`

In [70]:
step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

estimators = [
    ('rf', RandomForestRegressor(n_estimators=350,random_state=3,max_samples=0.5,max_features=0.75,max_depth=15)),
    ('gbdt',GradientBoostingRegressor(n_estimators=100,max_features=0.5)),
    ('xgb', XGBRegressor(n_estimators=25,learning_rate=0.3,max_depth=5))
]

step2 = StackingRegressor(estimators=estimators, final_estimator=Ridge(alpha=100))

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.8178892361081502
Mean Absolute error 2599.843806207983


So Gradient Boosting Regressor is performing well and giving highest accuracy in case of Validation Accurcay i.e of 82.51%.

### `Predicting the output of Test Data`

In [71]:
# using the pipeline of Gradient Boosting Regressor

step1 = ColumnTransformer(transformers=[
    ('scal_tnf',MinMaxScaler(),[0,2,3,5,6,7,8,11,12,13,15,16,19]),
    ('col_tnf',OneHotEncoder(sparse=False,drop='first'),[1,4,9,10,14,17,18])
],remainder='passthrough')

step2 = GradientBoostingRegressor(n_estimators=500)

pipe=Pipeline([
    ('step1',step1),
    ('step2',step2)
])

pipe.fit(X_train,y_train)
y_pred=pipe.predict(X_val)
print('R2 score',r2_score(y_val,y_pred))
print('Mean Absolute error',mean_absolute_error(y_val,y_pred))

R2 score 0.8248120202877873
Mean Absolute error 2576.762022920693


In [72]:
outputs_test = pipe.predict(X_test)

outputs_test

array([25250.91544352, 11904.95711865, 10458.83583862, ...,
       10017.00648537, 23659.22633595, 29828.06430381])

In [82]:
submission_df = pd.DataFrame({
    'rent':outputs_test
})

In [83]:
submission_df

rent
0     25250.915444
1     11904.957119
2     10458.835839
3      8817.129401
4     25489.736396
...            ...
4495  15750.192847
4496  13761.450987
4497  10017.006485
4498  23659.226336
4499  29828.064304

[4500 rows x 1 columns]

In [84]:
submission_df.to_csv('Submission.csv',index=None)

## `Loading the Pipeline and Dataset`

In [86]:
import pickle

pickle.dump(train_df,open('df.pkl','wb'))
pickle.dump(pipe,open('pipe.pkl','wb'))

Refrence : ![https://scikit-learn.org/stable/]